influx-reports/influx.py

```diff
@@ -0,0 +1,238 @@
+            client, measurement, job_reference, measurement_period_seconds)
+        report['runtime_seconds'] = measurement_period_seconds
+        report['per_field_uptime_percent'][measurement] = \
+            dict((measure, round(measure_uptime * 100, 2)) for
+                 measure, measure_uptime in
+                 measurement_data['per_field_uptime_percent'].items())
+    return dict(report)
+
+
+def main(args=None):
+    client = InfluxDBClient(args.influx_ip, args.influx_port,
+                            database='telegraf')
+    """
+    Influx structure:
+        Measurement (eg maas_nova)
+            Point (a timestamp with data)
+                Tags (Indexed K/V, eg job_reference)
+                Fields (Unindexed K/V, eg keystone_user_count)
+    """
+    # First find the first and last timestamp from telegraf.
+    # This way querying maas_* data will always
+    # be accurate (if no data is reported by maas plugins,
+    # this should be a failure metric)
+    find_first_timestamp_query = (
+        "select first(total) from processes "
+        "where job_reference = '{}';".format(args.job_reference)
+    )
+    first_ts = return_time(client, find_first_timestamp_query, delta_seconds=5)
+    find_last_timestamp_query = (
+        "select last(total) from processes "
+        "where job_reference = '{}';".format(args.job_reference))
+    last_ts = return_time(client, find_last_timestamp_query, delta_seconds=-5)
+
+    measurement_period_seconds = (last_ts - first_ts).total_seconds()
+
+    logging.info(
+        ("Metrics were gathered between {first} and {last}\n"
+         "For a total of {seconds} seconds".format(
+             first=first_ts,
+             last=last_ts,
+             seconds=measurement_period_seconds)))
+
+    measurements = ['maas_glance', 'maas_cinder', 'maas_keystone',
+                    'maas_heat', 'maas_neutron', 'maas_nova', 'maas_horizon']
+
+    report = build_report_dict(
+        client, measurements, args.job_reference, measurement_period_seconds)
+
+    if args.ymlreport:
+        yml_report = yaml.safe_dump(report, default_flow_style=False)
+        logging.info(yml_report)
+        with open(args.ymlreport, 'w+') as output_file:
+            output_file.write("---\n")
+            output_file.write(yml_report)
+
+    # Subunit is a unit test result format. Here we output a stream of
+    # test results, one per measurement.
+    # Each measurement is a successfull test if its overall uptime %age
+    # is higher than args.min_uptime
+    if args.subunitreport:
+        with SubunitContext(args.subunitreport) as output:
+            for measurement_name, measurement_data in \
+                    report['per_field_uptime_percent'].items():
+                for measure_name, uptime_percentage in \
+                        measurement_data.items():
+                    status = "fail"
+                    if uptime_percentage > args.max_downtime:
```

Comment:
This check is wrong. `uptime_percentage` would need to be less than 1 for the status to show a failure.